packages/signers/src/sign-transaction.ts

```diff
@@ -0,0 +1,202 @@
+    );
+
+    if (sendingSigner) {
+        const [signature] = await sendingSigner.signAndSendTransactions([signedTransaction]);
+        return signature;
+    }
+
+    if (fallbackSender) {
+        assertTransactionIsFullySigned(signedTransaction);
+        return fallbackSender(signedTransaction);
+    }
+
+    // TODO: Coded error.
+    throw new Error('No TransactionSendingSigner was identified and no fallback sender was provided.');
+}
+
+/**
+ * Identifies each provided TransactionSigner and categorizes them into their respective types.
+ * When a signer implements multiple interface, it will try to used to most powerful interface
+ * but fallback to the least powerful interface when necessary.
+ * For instance, if a signer implements TransactionSendingSigner and TransactionModifyingSigner,
+ * it will be categorized as a TransactionSendingSigner if and only if no other signers implement
+ * the TransactionSendingSigner interface.
+ */
+function categorizeTransactionSigners(
+    signers: readonly TransactionSigner[],
+    identifySendingSigner: boolean = true
+): Readonly<{
+    modifyingSigners: readonly TransactionModifyingSigner[];
+    partialSigners: readonly TransactionPartialSigner[];
+    sendingSigner: TransactionSendingSigner | null;
+}> {
+    // Identify the unique sending signer that should be used.
+    const sendingSigner = identifySendingSigner ? identifyTransactionSendingSigner(signers) : null;
+
+    // Now, focus on the other signers.
+    // I.e. the modifying or partial signers that are not the identified sending signer.
+    // Note that any other sending only signers will be discarded.
+    const otherSigners = signers.filter(
+        (signer): signer is TransactionModifyingSigner | TransactionPartialSigner =>
+            signer !== sendingSigner && (isTransactionModifyingSigner(signer) || isTransactionPartialSigner(signer))
+    );
+
+    // Identify the modifying signers from the other signers.
+    const modifyingSigners = identifyTransactionModifyingSigners(otherSigners);
+
+    // Use any remaining signers as partial signers.
+    const partialSigners = otherSigners
+        .filter(isTransactionPartialSigner)
+        .filter(signer => !(modifyingSigners as typeof otherSigners).includes(signer));
+
+    return Object.freeze({ modifyingSigners, partialSigners, sendingSigner });
+}
+
+/** Identifies the best signer to use as a TransactionSendingSigner, if any */
+function identifyTransactionSendingSigner(signers: readonly TransactionSigner[]): TransactionSendingSigner | null {
+    // Ensure there are any TransactionSendingSigners in the first place.
+    const sendingSigners = signers.filter(isTransactionSendingSigner);
+    if (sendingSigners.length === 0) return null;
+
+    // Prefer sending signers that do not offer other interfaces.
+    const sendingOnlySigners = sendingSigners.filter(
+        signer => !isTransactionModifyingSigner(signer) && !isTransactionPartialSigner(signer)
```

Comment:
Interesting choice. Wouldn't you expect that the most ‘potent’ signer be chosen (ie. the one that modifies _and_ sends)?

Reply:
I think there's a misunderstanding here.

No matter how many interfaces a signer implements, **only one of them will be used per address**. Providing multiple interfaces allows the app to decide which one to use.

Therefore, a signer that implements `TransactionModifyingSigner & TransactionSendingSigner` will either call `modifyAndSignTransactions` at the beginning OR `sendAndSendTransactions` at the end.

If a `TransactionSendingSigner` needs also needs to modify the transaction, they can do whatever they want since they are sending it anyway (as long as they don't invalidate previous signatures).

If a `TransactionSendingSigner` needs to modify the transaction whilst allowing others to partially sign it before it signs and sends it, then it simply need to offer a helper function that modify the transaction however it wishes. However a `modifyTransactions` function does not enter the realm of the Signer API as it does not sign anything.

Reply:
> Therefore, a signer that implements `TransactionModifyingSigner & TransactionSendingSigner` will either call `modifyAndSignTransactions` at the beginning OR `sendAndSendTransactions` at the end.

Ah, this brought the API more into focus for me. Thanks.